extract build progress from Ninja script output

christianrauch opened the pull request:
This extends the parsing of make stdout to get the build progress in the form of [XXX%] with parsing of Ninja stdout in the form of [X/Y]. Here, X and Y can be arbitrary large numbers with X<=Y, which is converted to a percentage.

christianrauch commented:
Not sure what the issue with the patch on the CI is, but this was probably introduced with https://github.com/colcon/colcon-notification/pull/61. This PR should be unrelated to this CI error, since it does not touch the patched file. Running `python3 setup.py --command-packages=stdeb.command bdist_deb` manually on my system (Ubuntu 22.04) creates the deb successfully without the `.app` folder.

cottsay commented:
> Not sure what the issue with the patch on the CI is

It's colcon/ci#10

christianrauch commented:
> > Not sure what the issue with the patch on the CI is
> 
> It's [colcon/ci#10](https://github.com/colcon/ci/pull/10)

Is there a fix for that? I see that the [`master`](https://github.com/colcon/colcon-notification/commits/master) branch has the same problem but this only appeared with the [most recent commit](https://github.com/colcon/colcon-notification/commit/8ae243584c7efd8d2912057d3911a6e156afcece).

Locally, this still seems to work and it is just an issue with the CI. Can we merge this nevertheless, given that the problem already exists on master, or is there going to be a fix on top of which I have to rebase my PR?

cottsay commented:
> Is there a fix for that?

That was the fix. It's merged now and I recovered both the `master` CI and this PR build.

christianrauch commented:
> > Is there a fix for that?
> 
> That was the fix. It's merged now and I recovered both the `master` CI and this PR build.

Sorry, I overlooked that this is in another repo. Thanks for fixing the CI.

Now that the CI passes, can you also review my PR? :smiley:

cottsay commented on colcon_notification/event_handler/status.py:
```diff
@@ -53,7 +53,9 @@ def __init__(self):  # noqa: D107
 
         # pattern to match progress indicator in e.g. make
-        self._progress_pattern = re.compile(r'^\[(  \d| \d\d|1\d\d)%\] ')
+        self._progress_pattern_make = re.compile(r'^\[(  \d| \d\d|1\d\d)%\] ')
+        # pattern to match progress indicator in ninja ([X/Y])
+        self._progress_pattern_ninja = re.compile(r'^\[([0-9]+)\/([0-9]+)\] ')
```
Can we generalize these variables? Maybe `_progress_pattern_percent` and `_progress_pattern_steps`? If/when we extend this for other tools we'll be able to simply update the existing patterns.

cottsay commented on colcon_notification/event_handler/status.py:
```diff
@@ -103,14 +105,19 @@ def __call__(self, event):  # noqa: D102
                 line = line.decode(errors='replace')
-            match = self._progress_pattern.match(line)
-            if not match:
+            match_make = self._progress_pattern_make.match(line)
+            match_ninja = self._progress_pattern_ninja.match(line)
+            if not match_make and not match_ninja:
```
I'd like to see this take a "first match wins" approach to avoid running `.match()` when we won't care.

cottsay commented on colcon_notification/event_handler/status.py:
```diff
@@ -103,14 +105,19 @@ def __call__(self, event):  # noqa: D102
             while len(progress) > 1:
                 progress.pop()
-            progress.append(match.group(1).lstrip() + '%')
+            if match_make:
+                progress.append(match_make.group(1).lstrip() + '%')
+            if match_ninja:
+                progress.append('{:.0%}'.format(
+                    int(match_ninja.group(1))/int(match_ninja.group(2))))
```
In the presence of both patterns (unlikely, but still), we should prefer the existing/more common pattern used by GNU make, which is not how this code will behave. Unless there is a specific reason to do the opposite, of course.

christianrauch commented:
Thanks for the review. I renamed the variables and changed the if/else conditions so that the "percentage" matching takes precedence and "fraction" matching is skipped if a "percentage" match has already been found.

cottsay reviewed (approved):
Thanks for your contribution.

I took the liberty of adding some simple tests here. Please take a look and tweak them as you see fit, and give me an @mention when you're satisfied.

christianrauch commented:
> I took the liberty of adding some simple tests here. Please take a look and tweak them as you see fit, and give me an @mention when you're satisfied.

@cottsay Thanks for adding the test. I played around with it a bit to make the test fail. It seems to behave as expected.
